### PR TITLE
ItemBlock model and phase 1 (single-thread) workflow changes

### PR DESCRIPTION
**Pull request by sseago**
Thank you for contributing to Velero!

# Please add a summary of your change

This PR adds the ItemBlock model and introduces the workflow changes for (single-threaded) ItemBlock functionality. ItemBlocks are created based on:
1) ItemBlockAction plugins for an item which return other related items are placed in the same ItemBlock
2) OrderedResources in the backup spec (ordered resources for a particular GroupResource are placed in the same ItemBlock to guarantee their relative backup order)
3) Items of the same GroupResource+Namespace+Name but different APIVersions (only happens if EnableAPIGroupVersions feature flag is set) are placed in the same ItemBlock

ItemBlocks are then backed up together using the new kubernetesBackupper.backupItemBlock func which
1) runs pre hooks for any pods in the ItemBlock
2) calls BackupItem on each item in the ItemBlock
3) runs post hooks for any pods in the ItemBlock

This change does not yet introduce worker threads to back up ItemBlocks in parallel -- that is out of scope for Velero 1.15
# Does your change fix a particular issue?

Third PR for #7900

# Please indicate you've done the following:

- [ x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ x] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Comment by sseago** [comment]
Converting to Draft for now until:
1) https://github.com/vmware-tanzu/velero/pull/8054 is merged
2) I get some more testing of actual backups and restores with this refactoring.

File: pkg/backup/backup.go [hunk 1]
```diff
@@ -501,8 +548,151 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
...
+				}
+			}
+		}
+		items = append(items, item)
+	}
+
+	itemBlock.Log.Debug("Backing up items in BackupItemBlock")
+	var grList []schema.GroupResource
+	for _, item := range items {
+		if backedUp := kb.backupItem(itemBlock.Log, item.Gr, itemBlock.itemBackupper, item.Item, item.PreferredGVR, &itemBlock); backedUp {
```
Comment (Lyndon-Li) [inline pkg/backup/backup.go:679]: For PVB, PVC is not to be backed up. How do we skip it since PVC and pod are in the same itemBlock?

File: pkg/backup/backup.go [hunk 2]
```diff
@@ -268,6 +275,12 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
 		return err
 	}
 
+	backupRequest.ResolvedItemBlockActions, err = itemBlockActionResolver.ResolveActions(kb.discoveryHelper, log)
+	if err != nil {
+		log.WithError(errors.WithStack(err)).Debugf("Error from itemBlockActionResolver.ResolveActions")
```
[the remark this answers is not in the document]
Comment (sseago) [reply pkg/backup/backup.go:280]: I followed the pattern already in place for resolving BIAs and getting resource hooks. But it does look odd. I could modify here, if it's ok to not follow the other pattern.

File: pkg/backup/backup.go [hunk 3]
```diff
@@ -402,37 +415,71 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
...
-		func() {
-			var unstructured unstructured.Unstructured
-
-			f, err := os.Open(item.path)
-			if err != nil {
-				log.WithError(errors.WithStack(err)).Error("Error opening file containing item")
-				return
+		// Skip if this item has already been added to an ItemBlock
+		if items[i].inItemBlock {
+			log.Infof("Skipping %s %s/%s because it's already in an ItemBlock", items[i].groupResource.String(), items[i].namespace, items[i].name)
```
[the remark this answers is not in the document]
Comment (sseago) [reply pkg/backup/backup.go:443]: I don't think it would be misunderstood as excluded, since "it's already in an ItemBlock", but I think it might make sense to make it debug. Any time an ItemBlock is created with multiple items, those items will later get this message.

Info makes it easier for users to see from the logs what's happening, though. I could go either way on this. But I'll modify the text a bit in either case. Maybe "Not creating new ItemBlock for  %s %s/%s because it's already in an ItemBlock"

File: pkg/backup/backup.go [hunk 4]
```diff
@@ -402,37 +415,71 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
...
+			}
+		}
 
-			if backedUp := kb.backupItem(log, item.groupResource, itemBackupper, &unstructured, item.preferredGVR); backedUp {
-				backedUpGroupResources[item.groupResource] = true
+		// We skip calling backupItemBlock here so that we will add the next item to the current ItemBlock if:
+		// 1) This is not the last item to be processed
+		// 2) Both current and next item are ordered resources
+		// 3) Both current and next item are for the same GroupResource
+		addNextToBlock := i < len(items)-1 && items[i].orderedResource && items[i+1].orderedResource && items[i].groupResource == items[i+1].groupResource
```
[the remark this answers is not in the document]
Comment (sseago) [reply pkg/backup/backup.go:474]: Yes. The item collector does one GR at a time, with all the items for a GR together, ordered items first. The intent here is to group all items that are "ordered resources" in a GR into one ItemBlock -- so if the current item is an ordered resource and the next item is also an ordered resource *and* the same GR, then it needs to be added to the same ItemBlock. If the next item is either not ordered or a different GR, then we're done with this block, and the next item will begin a new block.

File: pkg/backup/backup.go [hunk 5]
```diff
@@ -501,8 +548,151 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
...
+					namespace: metadata.GetNamespace(),
+					name:      metadata.GetName(),
+				}
+				// Don't run hooks if pod has already been backed up
+				if _, exists := itemBlock.itemBackupper.backupRequest.BackedUpItems[key]; !exists {
+					if err := itemBlock.itemBackupper.itemHookHandler.HandleHooks(itemBlock.Log, item.Gr, item.Item, itemBlock.itemBackupper.backupRequest.ResourceHooks, hook.PhasePre, itemBlock.itemBackupper.hookTracker); err != nil {
+						itemBlock.Log.WithError(err).WithField("name", item.Item.GetName()).Error("Error running hooks for pod")
+						// if pre hook fails, flag pod as backed-up and move on
+						itemBlock.itemBackupper.backupRequest.BackedUpItems[key] = struct{}{}
+						continue
```
[the remark this answers is not in the document]
Comment (sseago) [reply pkg/backup/backup.go:667]: The pod hook API's "fail on error" only calls for the pod itself to fail. Current Velero logic won't skip PVC, secrets, etc. either.

File: pkg/backup/backup.go [hunk 6]
```diff
@@ -501,8 +548,151 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
...
+			}
+			if !skipHooks {
+				key := itemKey{
+					resource:  resourceKey(item.Item),
+					namespace: metadata.GetNamespace(),
+					name:      metadata.GetName(),
+				}
+				// Don't run hooks if pod has already been backed up
+				if _, exists := itemBlock.itemBackupper.backupRequest.BackedUpItems[key]; !exists {
+					if err := itemBlock.itemBackupper.itemHookHandler.HandleHooks(itemBlock.Log, item.Gr, item.Item, itemBlock.itemBackupper.backupRequest.ResourceHooks, hook.PhasePre, itemBlock.itemBackupper.hookTracker); err != nil {
```
[the remark this answers is not in the document]
Comment (sseago) [reply pkg/backup/backup.go:663]: Sure, that makes sense. Generate a list here, then call a helper func and pass the list in which will (for now) just call HandleHooks sequentially and will return a status slice indicating which succeeded and which failed.

File: pkg/backup/backup.go [hunk 7]
```diff
@@ -501,8 +548,151 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
...
+	itemBlock.Log.Debug("Backing up items in BackupItemBlock")
+	var grList []schema.GroupResource
+	for _, item := range items {
+		if backedUp := kb.backupItem(itemBlock.Log, item.Gr, itemBlock.itemBackupper, item.Item, item.PreferredGVR, &itemBlock); backedUp {
+			grList = append(grList, item.Gr)
+		}
+	}
+
+	itemBlock.Log.Debug("Executing post hooks")
+	for _, pod := range postHookPods {
```
[the remark this answers is not in the document]
Comment (sseago) [reply pkg/backup/backup.go:685]: As before, we would pass the list to the helper, and in response get a list of errored pod hooks so we can handle them appropriately.

File: pkg/backup/backup.go (hunk 1 above)
Comment (sseago) [reply pkg/backup/backup.go:679]: But we do back up PVC. Current workflow backs up the item. It's in the item list. So we don't  want to skip it here.

File: pkg/backup/backup.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/backup/backup.go:280]: OK, if it doesn't looks good, we can modify the existing pattern, since we are doing refactors.

File: pkg/backup/backup.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/backup/backup.go:443]: I agree that this new message "Not creating new ItemBlock for %s %s/%s because it's already in an ItemBlock" is better.

And for the Info log vs. debug log, please decide it based on your own understanding, personally, I feel it more like a debug info.

File: pkg/backup/backup.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/backup/backup.go:667]: Understood. We do see some problems here though they also happen with the existing code:

- The hook may be used to quiescence the volume file system or prepare the backup data to the volume
- When the hook fails and the pod is skipped, but the PVC may be still backed up
- The consequence is that users either get a corrupted backup data or empty backup data

Do you think we can enhance this in future based on the new itemBlock implementation?

File: pkg/backup/backup.go (hunk 1 above)
Comment (Lyndon-Li) [reply pkg/backup/backup.go:679]: Understood. As the current logics, we reply on the handler of the PVC to decide whether to skip the PVC or not. See [this comment](https://github.com/vmware-tanzu/velero/pull/8102#discussion_r1737646097) for the problems I can see.

Let's discuss in that comment.

File: pkg/itemblock/itemblock.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+func (ib *ItemBlock) FindItem(gr schema.GroupResource, namespace, name string) []ItemBlockItem {
+	var itemList []ItemBlockItem
+	var returnList []ItemBlockItem
+
+	for _, item := range ib.Items {
+		if item.Gr == gr && item.Item != nil && item.Item.GetName() == name && item.Item.GetNamespace() == namespace {
+			itemGV, err := schema.ParseGroupVersion(item.Item.GetAPIVersion())
+			if err == nil && item.PreferredGVR.GroupVersion() == itemGV {
+				returnList = append(returnList, item)
+			} else {
```
Comment (blackpiglet) [inline pkg/itemblock/itemblock.go:54]: This logic does not look right.
The error branch will append the item even if the error was reported.

File: pkg/backup/item_backupper.go [hunk 9]
```diff
@@ -106,35 +107,21 @@ func (ib *itemBackupper) backupItem(logger logrus.FieldLogger, obj runtime.Unstr
...
-		"name":      name,
-		"resource":  groupResource.String(),
-		"namespace": namespace,
-	})
-
+func (ib *itemBackupper) itemInclusionChecks(log logrus.FieldLogger, mustInclude bool, metadata metav1.Object, obj runtime.Unstructured, groupResource schema.GroupResource) bool {
 	if mustInclude {
 		log.Infof("Skipping the exclusion checks for this resource")
 	} else {
 		if metadata.GetLabels()[velerov1api.ExcludeFromBackupLabel] == "true" {
```
Comment (blackpiglet) [inline pkg/backup/item_backupper.go:114]: It's better to check the GroupResource here.
Only the PVC and PV need to go through the PV skip tracking logic.
I know it's legacy code, but the change could save some condition checks.

**Comment by blackpiglet** [comment]
Maybe it's better to add a future enhancement, but it's better to know how the backed-up resources are organized with the ItemBlock, for example it can be displayed in the `velero backup describe --details` output.

File: pkg/backup/backup.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sseago) [reply pkg/backup/backup.go:667]: @Lyndon-Li I think we should defer this to the future, since we'll need to think carefully about the design. Do we *always* want to not restore items in the block that were returned by any IBA plugins for the pod if the pre-hook fails, or should this be part of the IBA plugin return value? If it's plugin-decided rather than optional, then we'll need to modify the IBA plugin return (with a v2 plugin). If we decide that we always want to skip related items for a pod if a hook fails, then we won't need that. But also, the model here doesn't track *which* item's IBA plugin (if any) was responsible for pulling the item into the block, so we'd need to add that too.

Also, this could introduce as many problems as it solves. If there are 2 pods in the block, either because of a shared RWX volume, or (in the future) due to the pods having volumes in the same VolumeGroup, then should one failed pod pre-hook cause all of the pods to be skipped in the backup? I think this should be a separate design discussion post-1.15.

[the remark this answers is not in the document]
Comment (sseago) [reply pkg/backup/backup.go:667]: It may be sufficient to just fail the pod -- the user will see the PartiallyFailed backup, note that the error was on a stateful pod and therefore assume the backup is bad anyway.

[the remark this answers is not in the document]
Comment (sseago) [reply pkg/backup/backup.go:667]: I would agree that if we decide to make a fix here in the future, it will be easier to do with the ItemBlock implementation than without.

File: pkg/itemblock/itemblock.go (hunk 8 above)
Comment (sseago) [reply pkg/itemblock/itemblock.go:54]: @blackpiglet this func returns a list of items matching the GR (could be more than one if EnableAPIGroupVersions is set). The purpose of the PreferredGVR comparison is that we want to return the item corresponding to the preferred GVR first, followed by the rest. So we loop over the list and when we find an item that's the preferred GVR, we add it to returnList, else we add to itemList -- then at the end, we append itemList to preferredList. So after filtering out the non-matching items (first `if`), there are 3 possibilities:
1)  We parse preferredGVR, compare it to item GVR and it's a match, so put the item at the front of the list (if branch)
2) We parse preferredGVR, compare it to item GVR and it's *not* a match, so put the item in the "other items" list (else branch)
3) We fail to parse preferredGVR, so we just put the item in the "other items" list (else branch).

Basically if we ever fail to parse preferred GVR, then we just return the list of items in the order we have them. If we parse it *and* there's a match (there should be), then that one goes first. We really don't expect this to ever fail, since the preferred GVR string was determined by the item collector, so there's no reason it would fail parsing -- but since ParseGroupVersion has the possibility of returning an error, we need to handle that unlikely case -- but we don't want to error out if it happened, since it would be a non-fatal error. Maybe I should add a debug log here in the case that error is not nil and separate that `if` from the comparison test to make the logic clearer, though.

File: pkg/backup/item_backupper.go (hunk 9 above)
Comment (sseago) [reply pkg/backup/item_backupper.go:114]: Hmm. So the tracking logic is only needed for PV/PC, but the exclusion checks overall apply to all types. i.e. any item is excludable, but the skipped PV tracking is only relevant for PVC/PV -- but that func already checks inside it. ExcludeFromBackupLabel, excluded namespace or resource type, etc. All of those checks are needed for even non-PV/PVC resources.

**Comment by sseago** [comment]
@blackpiglet Hmm. Yeah, we could definitely do that as a future enhancement. I guess we'd need to generate some sort of yaml file to upload to object storage -- basically when we generate an ItemBlock, before passing it on to start backup we could append yaml metadata, and at the end, we could upload this yaml to object store. I think it's fine to do in the  future, since it won't really affect much of this code and we're too late in the 1.15 cycle for new design, etc.

**Comment by sseago** [comment]
Updated PR based on feedback.

**Review by Lyndon-Li — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]
